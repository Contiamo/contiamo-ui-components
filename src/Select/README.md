The Select component presents users with a list of information with single-choice or multiple-choice options. Select elements can have options filled onClick, and also support filters.

It implements the specification for a **Collapsible Dropdown Listbox** according to [WAI-ARIA v1.1](https://www.w3.org/TR/wai-aria-practices/examples/listbox/listbox-collapsible.html).

### Basic usage

If the component is used with a string `value` prop, it behaves as a single select. Every time the value is changed, the options pop-up closes automatically. The `onChange` handler gives you the new value, and the option that was clicked as arguments. It is of the shape `(newValue: Value, changed: IOption) => void`.

```jsx
import * as React from "react"
import { Select, Code } from "@operational/components"

const options = [
  { label: "Option 1", value: "one" },
  { label: "Option 2", value: "two" },
  { label: "Option 3", value: "three" },
  { label: "Option 4", value: "four" },
  { label: "Option 5", value: "five" },
  { label: "Option 6", value: "six" },
  { label: "Option 7", value: "seven" },
  { label: "Option 8", value: "eight" },
]

const MyComponent = () => {
  const [value, setValue] = React.useState("one")
  const [lastChanged, setLastChanged] = React.useState(null)

  return (
    <>
      <Select
        id="basic"
        label="Basic"
        data-cy="basic-select"
        value={value}
        options={options}
        onChange={(newValue, lastChanged) => {
          setValue(newValue as string)
          setLastChanged(lastChanged)
        }}
      />
      <Code>
        {`
Current value: ${value}
Last changed option:

${JSON.stringify(lastChanged, null, 2)}
      `}
      </Code>
    </>
  )
}

;<MyComponent />
```

If options are provided without labels, values are be displayed in the list.

```jsx
import * as React from "react"
import { Select, Code } from "@operational/components"

const options = [
  { value: "one" },
  { value: "two" },
  { value: "three" },
  { value: "four" },
  { value: "five" },
  { value: "six" },
  { value: "seven" },
  { value: "eight" },
]

const MyComponent = () => {
  const [value, setValue] = React.useState("one")
  const [lastChanged, setLastChanged] = React.useState(null)

  return (
    <>
      <Select
        id="basic-no-labels"
        label="Without labels"
        data-cy="basic-select-no-labels"
        value={value}
        options={options}
        onChange={(newValue, lastChanged) => {
          setValue(newValue as string)
          setLastChanged(lastChanged)
        }}
      />
      <Code>
        {`
Current value: ${value}
Last changed option:

${JSON.stringify(lastChanged, null, 2)}
      `}
      </Code>
    </>
  )
}

;<MyComponent />
```

### Usage as multiselect

Using an array prop in the `value` makes the component work as multiselect. The pop-up stays open so that additional values may be added.

```jsx
import * as React from "react"
import { Select } from "@operational/components"

const options = [
  { label: "Option 1", value: "one" },
  { label: "Option 2", value: "two" },
  { label: "Option 3", value: "three" },
  { label: "Option 4", value: "four" },
  { label: "Option 5", value: "five" },
  { label: "Option 6", value: "six" },
  { label: "Option 7", value: "seven" },
  { label: "Option 8", value: "eight" },
]

const MyOtherComponent = () => {
  const [value, setValue] = React.useState([])
  return (
    <Select
      data-cy="multi-select"
      label="Select label"
      value={value}
      options={options}
      placeholder="Choose an option"
      onChange={newValue => setValue(newValue as string[])}
      filterable
    />
  )
}

;<MyOtherComponent />
```

### Filterable

Sometimes, if you have a lot of options, filtering through them might be helpful. Here's how you'd do that using a `filterable` prop.

```jsx
import * as React from "react"
import { Select } from "@operational/components"

const options = [
  { label: "Option 1", value: "one" },
  { label: "Option 2", value: "two" },
  { label: "Option 3", value: "three" },
  { label: "Option 4", value: "four" },
  { label: "Option 5", value: "five" },
  { label: "Option 6", value: "six" },
  { label: "Option 7", value: "seven" },
  { label: "Option 8", value: "eight" },
]

const MyThirdComponent = () => {
  const [value, setValue] = React.useState("one")
  return (
    <Select
      data-cy="filterable-select"
      label="Select label"
      value={value}
      options={options}
      filterable
      placeholder="Choose an option"
      onChange={newValue => setValue(newValue as string)}
    />
  )
}

;<MyThirdComponent />
```

### With `maxOptions`

If you have a huge list from a backend, you can limit the number of options displayed to avoid rendering performance issues. Just add `maxOptions` and it's done. Please note that without the `filterable` option enabled, some options can't be selected.

```jsx
import * as React from "react"
import { Select } from "@operational/components"

const options = [
  { label: "Option 1", value: "one" },
  { label: "Option 2", value: "two" },
  { label: "Option 3", value: "three" },
  { label: "Option 4", value: "four" },
  { label: "Option 5", value: "five" },
  { label: "Option 6", value: "six" },
  { label: "Option 7", value: "seven" },
  { label: "Option 8", value: "eight" },
]

const MyThirdComponent = () => {
  const [value, setValue] = React.useState("one")
  return (
    <Select
      data-cy="maxOptions-select"
      label="Select label"
      value={value}
      options={options}
      filterable
      maxOptions={2}
      placeholder="Choose an option"
      onChange={newValue => setValue(newValue as string)}
    />
  )
}

;<MyThirdComponent />
```

### Naked

Sometimes, you might have a need to render a `Select` without any surrounding chrome. The `naked` prop renders the component in a number of different contexts.

```jsx
import * as React from "react"
import { Select } from "@operational/components"

const options = [
  { label: "Option 1", value: "one" },
  { label: "Option 2", value: "two" },
  { label: "Option 3", value: "three" },
  { label: "Option 4", value: "four" },
  { label: "Option 5", value: "five" },
  { label: "Option 6", value: "six" },
  { label: "Option 7", value: "seven" },
  { label: "Option 8", value: "eight" },
]

const MyComponent = () => {
  const [value, setValue] = React.useState("one")

  return (
    <div style={{ width: "100%", display:"grid", gridGap: "10px"}}>
      <div style={{ width: "100%", background: "#666", display: "" }}>
        <Select
          value={value}
          options={options}
          filterable
          placeholder="Choose an option"
          onChange={newValue => setValue(newValue as string)}
        />
      </div>
      <div style={{ width: "100%", background: "#666", display: "" }}>
        <Select
          color="white"
          value={value}
          naked
          options={options}
          filterable
          placeholder="Choose an option"
          onChange={newValue => setValue(newValue as string)}
        />
      </div>
    </div>
  )
}

;<MyComponent />
```

### Disabled

In some cases, `Select`s must be disabled. We achieve this using the `disabled` prop.

```jsx
import * as React from "react"
import { Select } from "@operational/components"

const options = [
  { label: "Option 1", value: "one" },
  { label: "Option 2", value: "two" },
  { label: "Option 3", value: "three" },
  { label: "Option 4", value: "four" },
  { label: "Option 5", value: "five" },
  { label: "Option 6", value: "six" },
  { label: "Option 7", value: "seven" },
  { label: "Option 8", value: "eight" },
]

const MyComponent = () => {
  const [value, setValue] = React.useState("one")

  return (
    <Select
      data-cy="disabled-select"
      disabled
      value={value}
      options={options}
      label="Disabled"
      placeholder="Choose an option"
      onChange={newValue => setValue(newValue as string)}
    />
  )
}

;<MyComponent />
```

### With a Custom Input

In case of offering users a choice CSV delimiters, we have a few predefined ideas, but they can really be anything. In this case, let's offer them a custom option using the `customOption` property.

```jsx
import * as React from "react"
import { Select, Form, Code } from "@operational/components"

const options = [
  { label: "Option 1", value: "one" },
  { label: "Option 2", value: "two" },
  { label: "Option 3", value: "three" },
]

const MyComponent = () => {
  const [value, setValue] = React.useState("one")
  const [lastChanged, setLastChanged] = React.useState(null)
  const [customOptionValue, setCustomOptionValue] = React.useState("chickens")

  return (
    <Form>
      <Select
        data-cy="custom-select"
        value={value}
        options={options}
        label="Custom Input"
        placeholder="Choose an option"
        customOption={{ label: "Custom...", value: customOptionValue }}
        onChange={(newValue, lastChanged) => {
          setValue(newValue as string)
          if (lastChanged && lastChanged.label === "Custom...") {
            setCustomOptionValue(newValue as string)
          }
          setLastChanged(lastChanged)
        }}
      />
      <Code>
        {`
Current value: ${String(value)}
Last changed option:

${JSON.stringify(lastChanged, null, 2)}
      `}
      </Code>
    </Form>
  )
}

;<MyComponent />
```

### Return Value

The value prop passed to select is either an `Option` object, or an Array of `Option` objects. If it is an
array, the component automatically becomes a multi-select. The shape of an Option object is described below.

```ts
import { SelectProps } from "@operational/components"

const option: SelectProps["options"][-1] = {
  label: "any_string",
  value: "Any String",
}
```

### Custom Input exposes onBlur callback

Let's say we would like to prevent user leaving the custom input value empty, but still would like to allow user clean up the input before starting entering something new. We can use onInputBlur callback to implement such behaviour.

```jsx
import * as React from "react"
import { Select, Form, Code } from "@operational/components"

const options = [
  { label: "Option 1", value: "one" },
  { label: "Option 2", value: "two" },
  { label: "Option 3", value: "three" },
]

const MyComponent = () => {
  const [value, setValue] = React.useState("one")
  const [lastChanged, setLastChanged] = React.useState(null)
  const [customOptionValue, setCustomOptionValue] = React.useState("chickens")

  const onBlurHandler = value => {
    if (value === "") {
      setValue("chickens are back")
      setCustomOptionValue("chickens are back")
    }
  }

  return (
    <Form>
      <Select
        data-cy="custom-select"
        value={value}
        options={options}
        label="Custom Input"
        placeholder="Choose an option"
        customOption={{ label: "Custom...", value: customOptionValue }}
        onChange={(newValue, lastChanged) => {
          setValue(String(newValue))
          if (lastChanged && lastChanged.label === "Custom...") {
            setCustomOptionValue(String(newValue))
          }
          setLastChanged(lastChanged)
        }}
        customInputSettings={{
          onBlur: onBlurHandler,
        }}
      />
      <Code>
        {`
Current value: ${String(value)}
Last changed option:

${JSON.stringify(lastChanged, null, 2)}
      `}
      </Code>
    </Form>
  )
}

;<MyComponent />
```

### Very wide option

```jsx
import * as React from "react"
import { Select, Code } from "@operational/components"
const options = [
  { label: "Very looooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooong option", value: "one" },
  { label: "Short", value: "two" },
]
const MyComponent = () => {
  const [value, setValue] = React.useState("one")
  const [lastChanged, setLastChanged] = React.useState(null)
  return (
    <>
      <Select
        id="maxWidth"
        label="Max width"
        data-cy="maxWidth-select"
        value={value}
        options={options}
        onChange={(newValue, lastChanged) => {
          setValue(newValue as string)
          setLastChanged(lastChanged)
        }}
      />
      <Code>
        {`
Current value: ${value}
Last changed option:
${JSON.stringify(lastChanged, null, 2)}
      `}
      </Code>
    </>
  )
}
;<MyComponent />
```

### Select in Modal

```jsx
import * as React from "react"
import { Select, Code, Modal } from "@operational/components"
const options = [
  { label: "Very looooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooong option", value: "one" },
  { label: "Short", value: "two" },
]
const MyComponent = () => {
  const [value, setValue] = React.useState("one")
  const [lastChanged, setLastChanged] = React.useState(null)
  const [isModalOpen, setIsModalOpen] = React.useState(false)
  return (
    <>
      <div onClick={() => setIsModalOpen(true)}>Trigger the Modal</div>
      <Modal isOpen={isModalOpen} onClickOutside={() => setIsModalOpen(false)} title="What's up?">
        <Select
        id="modal"
        label="In modal"
        data-cy="inModal-select"
        value={value}
        options={options}
        onChange={(newValue, lastChanged) => {
          setValue(newValue as string)
          setLastChanged(lastChanged)
        }}
      />
      <Code>
        {`
Current value: ${value}
Last changed option:
${JSON.stringify(lastChanged, null, 2)}
      `}
      </Code>
      </Modal>
    </>
  )
}
;<MyComponent />
```
