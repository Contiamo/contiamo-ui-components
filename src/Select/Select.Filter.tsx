import * as React from "react"
import styled from "react-emotion"
import { OperationalStyleConstants } from "../utils/constants"

export interface Props {
  id?: string | number
  className?: string
  placeholder?: string
  onChange?: (newVal: string) => void
  color?: string
}

const Container = styled("div")(({ theme, color }: { theme?: OperationalStyleConstants; color?: string }) => {
  return {
    label: "selectfilter",
    "& > input": {
      width: "100%",
      padding: `${theme.deprecated.spacing / 2}px ${theme.deprecated.spacing * 3 / 4}px`,
      border: 0,
      outline: "none",
      font: "inherit",
    },
  }
})

const SelectFilter = (props: Props) => (
  <Container key={props.id} className={props.className}>
    <input
      onClick={e => e.stopPropagation()}
      onChange={(e: any) => {
        props.onChange(e.target.value)
      }}
      placeholder={props.placeholder || "Filter ..."}
    />
  </Container>
)

export default SelectFilter
