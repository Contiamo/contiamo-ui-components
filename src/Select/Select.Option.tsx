import * as React from "react"
import { darken, inputFocus } from "../utils"
import styled from "../utils/styled"
import { YesIcon } from "../Icon"

export interface Props {
  id?: number | string
  className?: string
  selected?: boolean
  onClick?: () => void
  children?: React.ReactNode
}

const Container = styled("div")<{ selected: boolean }>(({ theme, selected }) => {
  const backgroundColor = selected ? theme.color.background.lighter : theme.color.white
  return {
    backgroundColor,
    display: "flex",
    alignItems: "center",
    position: "relative",
    padding: `${theme.space.small}px ${theme.space.content}px`,
    wordWrap: "break-word",
    borderTop: "1px solid",
    borderColor: darken(backgroundColor, 10),
    color: theme.color.text.default,
    ":hover": {
      backgroundColor: darken(backgroundColor, 5),
    },
    ":focus": {
      ...inputFocus({ theme }),
    },
  }
})

const IconContainer = styled("div")(({ theme }) => {
  const size = 18
  return {
    display: "flex",
    alignItems: "center",
    flex: "1 0",
    flexBasis: size,
    justifyContent: "center",
    maxWidth: size,
    height: size,
    backgroundColor: theme.color.primary,
    borderRadius: "50%",
    marginLeft: "auto",
    "& svg": {
      color: theme.color.white,
      width: 12,
      height: 12,
    },
  }
})

const SelectOption = (props: Props) => (
  <Container
    key={props.id}
    className={props.className}
    selected={Boolean(props.selected)}
    tabIndex={-2}
    role="option"
    aria-selected={props.selected}
    onClick={(ev: React.SyntheticEvent<Node>) => {
      ev.stopPropagation()
      if (props.onClick) {
        props.onClick()
      }
    }}
  >
    {props.children}
    {props.selected ? (
      <IconContainer>
        <YesIcon size={10} />
      </IconContainer>
    ) : null}
  </Container>
)

export default SelectOption
