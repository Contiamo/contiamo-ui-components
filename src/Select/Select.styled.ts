import styled from "../utils/styled"
import { inputFocus } from "../utils"
import { expandColor } from "../utils/constants"
import Input from "../Input/Input"

export const Listbox = styled("div")<{ disabled: boolean; color?: string; fullWidth: boolean }>`
  label: SelectListbox;
  display: flex;
  flex-direction: column;
  color: ${({ theme, color }) => expandColor(theme, color)};
  cursor: ${({ disabled }) => (disabled ? "not-allowed" : "pointer")};
  opacity: ${({ disabled }) => (disabled ? 0.5 : 1)};
  outline: none;
  width: ${({ fullWidth }) => (fullWidth ? "100%" : "initial")};
`

const dropdownButtonWidth = 40

export const Combobox = styled("div")<{ naked: boolean; isOpen: boolean; hasCustomOption: boolean }>`
  display: grid;
  grid-template-columns: calc(100% - ${dropdownButtonWidth}px) ${dropdownButtonWidth}px;
  grid-gap: 1px;
  align-items: stretch;
  box-shadow: 0 0 0 1px
    ${({ theme, isOpen, hasCustomOption }) =>
      !hasCustomOption && isOpen ? theme.color.primary : theme.color.border.select};
  border-width: ${({ naked }) => (naked ? 0 : 1)}px;
  border-radius: ${({ theme }) => theme.borderRadius}px;
  background-color: ${({ naked }) => (naked ? "transparent" : "white")};

  :focus {
    ${({ theme }) => inputFocus({ theme })}
  }
`

export const SelectInput = styled(Input)<{ hasCustomOption: boolean }>`
  width: inherit;
  border: 0;
  background: transparent;
  color: currentColor;
  cursor: inherit;
  border-top-right-radius: 0;
  border-bottom-right-radius: 0;
  overflow: hidden;
  text-overflow: ellipsis;
  box-shadow: none;
  :focus {
    box-shadow: none;
  }
`

export const FilterInput = styled(Input)`
  border: 0;
  margin: ${({ theme }) => -theme.space.content}px;
  border-radius: 0;
  height: auto;
  min-height: 32px;
  max-width: none;
`

export const DropdownButton = styled("div")<{ naked: boolean; isOpen: boolean; hasCustomOption: boolean }>`
  display: flex;
  align-items: center;
  justify-content: center;
  margin: -1px 0;
  border: none;
  border-top-right-radius: ${({ theme }) => theme.borderRadius}px;
  border-bottom-right-radius: ${({ theme }) => theme.borderRadius}px;

  :focus {
    box-shadow: none;
  }

  ::after {
    content: "";
    width: 0;
    height: 0;
    border: 4px solid transparent;
    border-top-color: ${({ theme, isOpen, naked }) =>
      !naked ? (isOpen ? theme.color.primary : theme.color.text.default) : "currentColor"};
    transform: ${({ isOpen }) => (isOpen ? `translateY(-2px) rotate(180deg)` : `translateY(2px)`)};
  }
`

DropdownButton.defaultProps = { role: "button", "aria-disabled": false, "aria-label": "Expand" }
