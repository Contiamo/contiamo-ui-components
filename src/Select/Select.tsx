import * as React from "react"
import { DefaultProps } from "../types"
import styled from "../utils/styled"

import { height as inputHeight } from "../Input/Input.constants"
import { LabelText } from "../LabelText/LabelText"
import { floatIn, keyCodes, readableTextColor, resetTransform } from "../utils"
import { expandColor } from "../utils/constants"
import { inputFocus, Label } from "../utils/mixins"
import SelectFilter from "./Select.Filter"
import SelectOption from "./Select.Option"

export type Value = number | string

export interface IOption {
  label?: string
  value: Value
}

const displayOption = (opt: IOption): string => {
  if (opt.label) {
    return opt.label
  }

  return String(opt.value)
}

export interface SelectProps extends DefaultProps {
  /** Options available */
  options: IOption[]
  /** Current value */
  value: null | Value | Value[]
  /** Make the list filterable */
  filterable?: boolean
  /**
   * Limit the number of options displayed
   */
  maxOptions?: number
  /** Disable the component */
  disabled?: boolean
  /** Callback trigger on any changes */
  onChange?: (newValue: null | Value | Value[], changedItem?: Value) => void
  /** Text color */
  color?: string
  /** Text to display when no active selection */
  placeholder?: string
  /** Label text */
  label?: string
  /** Should the Select be rendered with a full box style? */
  naked?: boolean
}

export interface State {
  open: boolean
  updating: boolean
  search: string
}

const Container = styled("div")<Partial<SelectProps>>(({ theme, color, disabled, naked }) => {
  const backgroundColor = naked ? "transparent" : expandColor(theme, color) || theme.color.white
  const dropdownArrowWidth = 56
  return {
    backgroundColor,
    label: "select",
    position: "relative",
    display: "flex",
    alignItems: "center",
    padding: `${theme.space.small}px ${dropdownArrowWidth}px ${theme.space.small}px ${theme.space.content}px`,
    borderRadius: 4,
    width: "fit-content",
    minWidth: !naked ? 240 : "none",
    minHeight: inputHeight,
    border: naked ? 0 : "1px solid",
    borderColor: theme.color.border.default,
    opacity: disabled ? 0.5 : 1,
    cursor: "pointer",
    color: readableTextColor(backgroundColor, ["black", "white"]),
    outline: "none",
    pointerEvents: disabled ? "none" : "all",
    // downward caret.
    "&::after": {
      content: "''",
      position: "absolute",
      top: "50%",
      right: theme.space.small,
      width: 0,
      height: 0,
      border: "4px solid transparent",
      borderTopColor: theme.color.border.default,
      transform: "translateY(calc(-50% + 2px))",
    },
    "&:focus":
      !naked &&
      inputFocus({
        theme,
      }),
  }
})

const DisplayValue = styled("div")<{ isPlaceholder: boolean }>(({ theme, isPlaceholder }) => {
  if (isPlaceholder) {
    return {
      color: theme.color.text.lightest,
    }
  }

  return {
    color: "currentColor",
  }
})

const Options = styled("div")(
  {
    position: "absolute",
    /**
     * Push it down 6px so it doesn't overlap with the focus shadow,
     * and there's a comfortable 2px gap.
     */
    top: "calc(100% + 6px)",
    left: 0,
    minWidth: "100%",
    overflow: "hidden",
    borderRadius: 4,
    opacity: 0,
    transform: "translateY(-10px)",
    animation: `${floatIn} .15s forwards ease,
    ${resetTransform} .15s forwards ease`,
  },
  ({ theme }) => ({
    boxShadow: "0 3px 12px rgba(0, 0, 0, .14)",
    zIndex: theme.zIndex.selectOptions,
  }),
)

const OptionsList = styled("div")({
  maxHeight: 200,
  overflow: "auto",
})

class Select extends React.Component<SelectProps, State> {
  public state: State = {
    open: false,
    updating: false,
    search: "",
  }

  public containerNode?: Node

  public static defaultProps: Partial<SelectProps> = {
    placeholder: "No entries selected",
    naked: false,
  }

  // This implements "click outside to close" behavior
  public handleClick = (ev: MouseEvent): void => {
    // if we're clicking on the Select itself,
    if (this.containerNode && this.containerNode.contains(ev.target as Node)) {
      return
    }

    // if we're clicking outside,
    this.close()
  }

  public handleEsc = (e: KeyboardEvent) => {
    if (e.keyCode === keyCodes.esc) {
      this.close()
    }
  }

  public componentDidUpdate() {
    if (this.state.open) {
      document.addEventListener("click", this.handleClick)
      document.addEventListener("keyup", this.handleEsc)
    } else {
      document.removeEventListener("click", this.handleClick)
      document.removeEventListener("keyup", this.handleEsc)
    }
  }

  public getDisplayValue(): string | undefined {
    const { placeholder } = this.props

    if (!this.props.value) {
      return placeholder
    }

    if (!Array.isArray(this.props.value)) {
      const displayedOption = this.props.options.filter(option => option.value === this.props.value)[0]
      return displayedOption ? displayOption(displayedOption) : placeholder
    }

    const listDisplay = this.props.options
      .map(option => ((this.props.value as Value[]).indexOf(option.value) > -1 ? displayOption(option) : null))
      .filter(a => !!a)
      .join(", ")
    return listDisplay === "" ? this.props.placeholder || "" : listDisplay
  }

  public selectOption(option: IOption) {
    const { onChange } = this.props

    if (!onChange) {
      return
    }

    if (!Array.isArray(this.props.value)) {
      this.setState({
        open: false,
      })
      onChange(this.props.value === option.value ? null : option.value)
      return
    }

    const optionIndex: number = this.props.value.indexOf(option.value)

    if (optionIndex < 0) {
      onChange([...this.props.value, option.value], option.value)
    } else {
      onChange([...this.props.value.slice(0, optionIndex), ...this.props.value.slice(optionIndex + 1)], option.value)
    }
  }

  public isOptionSelected(option: IOption) {
    if (!Array.isArray(this.props.value)) {
      return this.props.value === option.value
    }

    return this.props.value.indexOf(option.value) > -1
  }

  public close() {
    this.setState(() => ({
      open: false,
    }))
  }

  public render() {
    const { color, disabled, naked, value, options, filterable, label, onChange, maxOptions, ...props } = this.props
    const { open, search } = this.state
    const selectWithoutLabel = (
      <Container
        {...props}
        color={color}
        disabled={disabled}
        naked={naked}
        innerRef={(containerNode: HTMLElement) => (this.containerNode = containerNode)}
        role="listbox"
        tabIndex={-2}
        onClick={() => {
          this.setState(prevState => ({
            open: !prevState.open,
          }))
        }}
      >
        <DisplayValue isPlaceholder={Array.isArray(value) ? value.length === 0 : !value}>
          {this.getDisplayValue()}
        </DisplayValue>
        {Boolean(options.length) &&
          open && (
            <Options>
              {filterable && (
                <SelectFilter
                  onChange={(filterValue: string) => {
                    this.setState({
                      search: filterValue,
                    })
                  }}
                />
              )}
              <OptionsList>
                {options
                  .filter(option => (option.label || String(option.value)).match(RegExp(search, "i")))
                  .map(option => (
                    <SelectOption
                      key={String(option.value)}
                      onClick={() => {
                        this.selectOption(option)
                      }}
                      selected={this.isOptionSelected(option)}
                    >
                      {option.label || String(option.value)}
                    </SelectOption>
                  ))
                  .slice(0, maxOptions)}
              </OptionsList>
            </Options>
          )}
      </Container>
    )
    return label ? (
      <Label {...props}>
        <LabelText>{label}</LabelText>
        {selectWithoutLabel}
      </Label>
    ) : (
      selectWithoutLabel
    )
  }
}

export default Select
