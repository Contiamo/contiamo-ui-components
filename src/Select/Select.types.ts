import { DefaultProps } from "../types"

export type Value = number | symbol | string

export interface IOption {
  label?: string
  value: Value
}

export interface BaseSelectProps extends DefaultProps {
  /** Options available */
  options: IOption[]
  /** Make the list filterable */
  filterable?: boolean
  /** Limit the number of options displayed */
  maxOptions?: number
  /** Disable the component */
  disabled?: boolean
  /** Callback trigger on any changes */
  onChange?: (newValue: null | Value | Value[], changedItem?: IOption) => void
  /** Text color */
  color?: string
  /** Text to display when no active selection */
  placeholder?: string
  /** Label text */
  label?: string
  /** Should the Select be rendered with a full box style? */
  naked?: boolean
  /** We never have children */
  children?: never
  /** Shall the component be full width? */
  fullWidth?: boolean
}

export interface SelectPropsWithCustomOption extends BaseSelectProps {
  /** A pseudo-option that allows a user to supply a custom value. */
  customOption?: IOption
  /** Current value */
  value: null | Value
}

export interface SelectPropsWithMultiSelect extends BaseSelectProps {
  /** We cannot have a custom option in a multi-selectable Select component */
  customOption?: never
  /** The value of the pseudo-option that allows a user to supply a custom value. */
  customOptionValue?: never
  /** Current value */
  value: null | Value | Value[]
}

export type SelectProps = SelectPropsWithCustomOption | SelectPropsWithMultiSelect
