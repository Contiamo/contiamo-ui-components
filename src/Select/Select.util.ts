import { IOption, SelectProps, Value } from "./Select.types"
import { IContextMenuItem } from "../ContextMenu/ContextMenu.Item"

export const appendItem = (newItem: IContextMenuItem) => (existingItems: IContextMenuItem[]): IContextMenuItem[] => [
  ...existingItems,
  newItem,
]

export const truncateList = (maxLength?: number) => (options: SelectProps["options"]) =>
  maxLength ? options.slice(0, maxLength) : options

export const filterList = (filter: string) => {
  const searchRegExp = new RegExp(filter.replace(/[.*+?^${}()|[\]\\]/g, "\\$&"), "ig")
  return (options: SelectProps["options"]) => options.filter(option => String(option.label).match(searchRegExp))
}

export const prependItem = (filterItem: IContextMenuItem) => (items: IContextMenuItem[]): IContextMenuItem[] => [
  filterItem,
  ...items,
]

export const isOptionSelected = (value: Value | Value[] | null) => (option: IOption) => {
  if (!Array.isArray(value)) {
    return value === option.value
  }

  return value.includes(option.value)
}

export const getNewValue = (value: Value | Value[] | null) => (newValue: Value) => {
  if (Array.isArray(value)) {
    const newValueIndex = value.indexOf(newValue)
    if (newValueIndex === -1) {
      return [...value, newValue]
    } else {
      return [...value.slice(0, newValueIndex), ...value.slice(newValueIndex + 1)]
    }
  }
  return newValue
}

export const getDisplayValue = (internalValue: Value | Value[] | null) => (options: IOption[]): string => {
  if (Array.isArray(internalValue)) {
    return options
      .filter(option => internalValue.includes(option.value))
      .map(option => option.label)
      .join(", ")
  }

  const valueExistsInOptions = options.find(option => option.value === internalValue)
  if (valueExistsInOptions) {
    return valueExistsInOptions.label || String(valueExistsInOptions.value)
  }

  if (internalValue === null) {
    return ""
  }

  return String(internalValue)
}

export const optionsToContextMenuItems = (overrides?: (option: IOption) => Partial<IContextMenuItem>) => (
  options: IOption[],
) =>
  options.map(
    (option): IContextMenuItem => ({
      ...option,
      label: option.label || String(option.value), // fall back to the value if no label
      ...(overrides ? overrides(option) : {}),
    }),
  )

export const getOptionFromItem = (haystack: IContextMenuItem[]) => (needle: IContextMenuItem): IOption | undefined => {
  const result = haystack.find(item => needle.label === item.label)

  if (result) {
    return { label: result.label || result.value, value: result.value }
  }

  return undefined
}
