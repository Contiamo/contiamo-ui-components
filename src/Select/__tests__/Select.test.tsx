import {
  appendItem,
  filterList,
  getDisplayValue,
  getNewValue,
  isOptionSelected,
  optionsToContextMenuItems,
  prependItem,
  truncateList,
} from "../Select.util"
import { IOption } from "../Select.types"

describe("Select component utilities", () => {
  const defaultOptionValue = "🇪🇸🌍 queso"
  const myOptions: IOption[] = [{ label: "Hola Mundo", value: defaultOptionValue }]
  const longTestOptions: IOption[] = [
    ...Array.from({ length: 100 }, (_, index) => ({ label: String(index), value: index })),
    ...myOptions,
    ...myOptions,
    ...myOptions,
  ]

  it("should convert options to context menu items", () => {
    const mockFn = jest.fn
    expect(optionsToContextMenuItems(() => ({ onClick: mockFn }))(myOptions)).toEqual([
      { label: "Hola Mundo", value: defaultOptionValue, onClick: mockFn },
    ])
  })

  it("should prepend an option to an existing list of options", () => {
    expect(prependItem({ icon: "Add", label: "Sup" })(optionsToContextMenuItems()(myOptions))).toEqual([
      { icon: "Add", label: "Sup" },
      { label: "Hola Mundo", value: defaultOptionValue },
    ])
  })
  it("should append an option to an existing list of options", () => {
    expect(appendItem({ icon: "Add", label: "Sup" })(optionsToContextMenuItems()(myOptions))).toEqual([
      { label: "Hola Mundo", value: defaultOptionValue },
      { icon: "Add", label: "Sup" },
    ])
  })
  it("should filter an existing list of options based on their label", () => {
    expect(filterList("hol")(longTestOptions)).toEqual([
      { label: "Hola Mundo", value: defaultOptionValue },
      { label: "Hola Mundo", value: defaultOptionValue },
      { label: "Hola Mundo", value: defaultOptionValue },
    ])
  })
  it("should truncate a list of options", () => {
    expect(truncateList(2)(longTestOptions)).toEqual([{ label: "0", value: 0 }, { label: "1", value: 1 }])
  })
  it("should truncate a list of options while respecting filters", () => {
    expect(truncateList(2)(filterList("hol")(longTestOptions))).toEqual([
      { label: "Hola Mundo", value: defaultOptionValue },
      { label: "Hola Mundo", value: defaultOptionValue },
    ])
  })
  it("should get a display value from a custom input value", () => {
    const customInputSymbol = Symbol("lol")
    expect(getDisplayValue(customInputSymbol, "hello", customInputSymbol)).toEqual("hello")
  })
  it("should get a display value from a multi-select value", () => {
    expect(getDisplayValue(["my name is earl", "what's your name", "follow @hackSultan"])).toEqual(
      "my name is earl, what's your name, follow @hackSultan",
    )
  })
  it("should get a display value from a value that is null or not yet initialized", () => {
    expect(getDisplayValue(null)).toEqual("")
  })
  it("should get a display value from a primitive string value", () => {
    expect(getDisplayValue("cheese is my friend")).toEqual("cheese is my friend")
  })
  it("should get a display value from a numeric value", () => {
    expect(getDisplayValue(4)).toEqual("4")
  })
  it("should get a changed value when a basic value changes", () => {
    expect(getNewValue("cheese")("hola")).toEqual("hola")
  })
  it("should get a correct changed value when a multi-select option is added", () => {
    expect(getNewValue(["Voss", "rocks"])("Laurie")).toEqual(["Voss", "rocks", "Laurie"])
  })
  it("should get a correct changed value when a multi-select option is removed", () => {
    expect(getNewValue(["Laurie", "Voss", "rocks"])("Laurie")).toEqual(["Voss", "rocks"])
  })
  it("should be able to tell if an option is selected", () => {
    expect(isOptionSelected("chickens")({ label: "I am not a chicken", value: "what's up" })).toEqual(false)
    expect(isOptionSelected("chickens")({ label: "I am not a chicken", value: "chickens" })).toEqual(true)
  })
  it("should be able to tell if multi-select options are selected", () => {
    expect(isOptionSelected(["chickens", "cheese"])({ label: "I am not a chicken", value: "what's up" })).toEqual(false)
    expect(isOptionSelected(["chickens", "cheese"])({ label: "I am not a chicken", value: "chickens" })).toEqual(true)
  })
})
