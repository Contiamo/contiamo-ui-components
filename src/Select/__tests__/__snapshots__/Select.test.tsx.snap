// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select Should render correctly 1`] = `
<div
  class="css-ftf61z"
>
  <div
    class="css-p29iif"
  />
  <div
    class="css-ag9c3l-select"
    placeholder="Select me"
    role="listbox"
    tabindex="-2"
  >
    <div
      class="css-thzv76"
    >
      Select me
    </div>
  </div>
</div>
`;
